lib/pbench/server/filetree.py

```diff
@@ -298,20 +327,84 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception/failure occurs while unpacking
+            the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
-        unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
-        self.unpacked = unpacked
+        unpacked = incoming / f"{self.name}.unpack"
+        unpacked.mkdir(parents=True)
+
+        try:
+            process = subprocess.run(
+                [
+                    "tar",
+                    "-x",
+                    "--no-same-owner",
+                    "--delay-directory-restore",
+                    "--force-local",
+                    "--file=" + str(self.tarball_path),
+                ],
+                cwd=unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
+        except Exception as exc:
+            raise TarballUnpackError(self.tarball_path, str(exc)) from exc
+        else:
+            if process.returncode != 0:
+                raise TarballUnpackError(
+                    self.tarball_path,
+                    f"tar exited with status {process.returncode}:  {process.stderr.strip()!r}",
+                )
+
+        find_command = "find . ( -type d -exec chmod ugo+rx '{}' + ) -o ( -type f -exec chmod ugo+r '{}' + )"
+        try:
+            process = subprocess.run(
+                find_command.split(),
+                cwd=unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
+        except Exception as exc:
+            raise TarballModeChangeError(unpacked, str(exc)) from exc
+        else:
+            if process.returncode != 0:
+                raise TarballModeChangeError(
+                    unpacked,
+                    f"tar exited with status {process.returncode}:  {process.stderr.strip()!r}",
+                )
+
+        try:
+            shutil.move(unpacked / self.name, incoming)
+        except Exception as exc:
+            try:
+                shutil.rmtree(incoming)
```

Comment:
`incoming` is the *controller* directory, `/srv/pbench/public_html/<controller>`. We do *not* want to remove that, which includes all other unpacked tarballs!

We probably want to attempt here to remove `incoming / self.name` in case we created the directory before failure.